src/main/java/org/logstash/plugins/inputs/http/util/ExecutionObserver.java

```diff
@@ -0,0 +1,201 @@
+
+        Stats(int nodes, int executing) {
+            this.nodes = nodes;
+            this.executing = executing;
+        }
+    }
+
+    @FunctionalInterface
+    public interface ExceptionalSupplier<T, E extends Throwable> {
+        T get() throws E;
+    }
+
+    public <T,E extends Throwable> T observeExecution(final ExceptionalSupplier<T,E> supplier) throws E {
+        final Execution execution = startExecution();
+        try {
+            return supplier.get();
+        } finally {
+            final boolean isCompact = execution.markComplete();
+            if (!isCompact) {
+                this.compactTail();
+            }
+        }
+    }
+
+    @FunctionalInterface
+    public interface ExceptionalRunnable<E extends Throwable> {
+        void run() throws E;
+    }
+
+    public <E extends Throwable> void observeExecution(final ExceptionalRunnable<E> runnable) throws E {
+        observeExecution(() -> { runnable.run(); return null; });
+    }
+
+    // visible for test
+    Execution startExecution() {
+        final Execution newHead = new Execution(nanosSupplier.getAsLong());
+
+        // atomically attach the new execution as a new (detached) head
+        final Execution oldHead = this.head.getAndSet(newHead);
+        // attach our new head to the old one
+        oldHead.linkNext(newHead);
+
+        return newHead;
+    }
+
+    private Execution compactTail() {
+        return this.tail.updateAndGet(Execution::chaseTail);
+    }
+
+    static class Execution {
+        private static final java.lang.invoke.VarHandle NEXT;
+        static {
+            try {
+                MethodHandles.Lookup l = MethodHandles.lookup();
+                NEXT = l.findVarHandle(Execution.class, "next", Execution.class);
+            } catch (ReflectiveOperationException e) {
+                throw new ExceptionInInitializerError(e);
+            }
+        }
+
+        private final long startNanos;
+
+        private volatile boolean isComplete;
+        private volatile Execution next;
+
+        Execution(long startNanos) {
+            this(startNanos, false);
+        }
+
+        Execution(final long startNanos,
+                  final boolean isComplete) {
+            this.startNanos = startNanos;
+            this.isComplete = isComplete;
+        }
+
+        /**
+         * marks this execution as complete
+         * @return true if the completion resulted in a compaction
+         */
+        boolean markComplete() {
+            isComplete = true;
+
+            // concurrency: use plain memory for reads because we can tolerate
+            // completed nodes remaining as the result of a race
+            final Execution preCompletionNext = this.getNextPlain();
+            if (preCompletionNext != null) {
+                final Execution newNext = preCompletionNext.chaseTail();
+                return (newNext != preCompletionNext) && NEXT.compareAndSet(this, preCompletionNext, newNext);
+            }
```

Comment:
review note:

We can't actually compact the completed node out, since we don't have a reference to its prev and a doubly-linked list is more complex than we need.

But we _can_ fast-forward a newly-completed node's `next` pointer to the first node in the sequence that is either not yet completed or has no next, which may be a compaction that frees up memory and reduces future computation.

When we succeed at compacting, we return true so that the caller can make decisions about performing additional compactions.